src/gather_classes.rs

```diff
@@ -257,80 +255,131 @@ fn update_report(
-                Ok(Box::new(::std::iter::empty()))
-            } else {
-                Err(anyhow!("unexpected string: {}", string))
+    fn get_calls_inner(obj: &Value, result: &mut Vec<CallWithCount>) -> Result<(), anyhow::Error> {
+        match obj {
+            Value::String(string) => {
+                // If the string is SAME or EMPTY, then there are no calls to extract and continue processing
+                if string_is_same(string) || string_is_empty(string) {
+                    Ok(())
+                } else {
+                    Err(anyhow!("unexpected string: {}", string))
+                }
             }
-        }
-        Value::Array(call_list) => {
-            let nested_calls = call_list
-                .iter()
-                .map(|inner_call| get_calls(inner_call))
-                .collect::<Result<Vec<_>, _>>()?
-                .into_iter()
-                .flatten();
-            Ok(Box::new(call_list.iter().chain(nested_calls)))
-        }
-        Value::Object(obj_map) => {
-            if let Some(Value::Object(diff_map)) = obj_map.get(DIFFERENT) {
-                let base_list = diff_map
-                    .get("base")
-                    .ok_or(anyhow!("difference should have base"))?;
-                let native_list = diff_map
-                    .get("native")
-                    .ok_or(anyhow!("difference should have native"))?;
-                // TODO (#72) Compare and get diffs. For now we just chain.
-                Ok(Box::new(
-                    get_calls(base_list)?.chain(get_calls(native_list)?),
-                ))
-            } else if let Some(calls_value) = obj_map.get("calls") {
-                get_calls(calls_value)
-            } else if let Some(_val) = obj_map.get("revert_reason") {
-                Ok(Box::new(::std::iter::empty()))
-            } else {
-                Err(anyhow!("unexpected object: `{}`", obj))
+            Value::Array(call_list) => {
+                for call in call_list {
+                    get_calls_inner(call, result)?;
+                }
+                Ok(())
+            }
+            Value::Object(obj_map) => {
+                // If call is different, then base and native will have their own list of calls
+                if let Some(Value::Object(diff_map)) = obj_map.get(DIFFERENT) {
+                    let base_list = diff_map
+                        .get("base")
+                        .ok_or(anyhow!("difference should have base"))?;
+                    let native_list = diff_map
+                        .get("native")
+                        .ok_or(anyhow!("difference should have native"))?;
+
+                    let mut base_calls = Vec::new();
+                    get_calls_inner(base_list, &mut base_calls)?;
+                    let mut native_calls = Vec::new();
+                    get_calls_inner(native_list, &mut native_calls)?;
+
+                    result.extend(merge_calls_with_count(base_calls, native_calls));
```

Comment:
This implementation doesn't really consider the different layers for the calls. Right now this would naively just compare 2 dictionaries and return `{ call_1: 1, call_3: 1 }`.

Do you think call 2 and call 3 should be removed entirely since they are of different depths?

Reply:
I will leave it as another ticket to resolve #91